src/Features/Core/Portable/Completion/CompletionService.ProviderManager.cs

```diff
@@ -57,41 +54,74 @@ public ProviderManager(CompletionService service)
             {
                 roles ??= ImmutableHashSet<string>.Empty;
 
                 lock (_gate)
                 {
-                    return _rolesToProviders.GetOrAdd(roles, _createRoleProviders);
+                    if (!_rolesToProviders.TryGetValue(roles, out var providers))
+                    {
+                        providers = GetImportedAndBuiltInProvidersWorker(roles);
+                        _rolesToProviders.Add(roles, providers);
+                    }
+
+                    return providers;
+                }
+
+                ImmutableArray<CompletionProvider> GetImportedAndBuiltInProvidersWorker(ImmutableHashSet<string> roles)
+                {
+                    using var _ = ArrayBuilder<CompletionProvider>.GetInstance(out var providers);
+                    providers.AddRange(GetLazyImportedProviders()
+                        .Where(lz => lz.Metadata.Roles == null || lz.Metadata.Roles.Length == 0 || roles.Overlaps(lz.Metadata.Roles))
+                        .Select(lz => lz.Value));
+
+#pragma warning disable 0618
+                    // We need to keep supporting built-in providers for a while longer since this is a public API.
+                    // https://github.com/dotnet/roslyn/issues/42367
+                    providers.AddRange(_service.GetBuiltInProviders());
+#pragma warning restore 0618
+
+                    return providers.ToImmutable();
                 }
             }
 
-            public CompletionProvider? GetProvider(CompletionItem item)
+            public ImmutableArray<CompletionProvider> GetProviders(ImmutableHashSet<string> roles, Project? project)
+            {
+                using var _ = ArrayBuilder<CompletionProvider>.GetInstance(out var providers);
+                providers.AddRange(GetImportedAndBuiltInProviders(roles));
+                providers.AddRange(GetProjectCompletionProviders(project));
+                return providers.ToImmutable();
+            }
+
+            public CompletionProvider? GetProvider(CompletionItem item, Project? project)
             {
                 CompletionProvider? provider = null;
 
                 if (item.ProviderName != null)
                 {
                     lock (_gate)
                     {
-                        provider = _nameToProvider.GetOrAdd(item.ProviderName, _getProviderByName);
+                        if (!_nameToProvider.TryGetValue(item.ProviderName, out provider))
+                        {
+                            provider = GetImportedAndBuiltInProviders(roles: ImmutableHashSet<string>.Empty).FirstOrDefault(p => p.Name == item.ProviderName);
+                            _nameToProvider.Add(item.ProviderName, provider);
+                        }
                     }
+
+                    provider ??= GetProjectCompletionProviders(project).FirstOrDefault(p => p.Name == item.ProviderName);
```

Comment:
This is the core to the fix, we didn't consult project references for operations tied to completion item befire (get description, text change, etc.)

Reply:
Should we check if project is null to avoid any unnecessary work?

Reply:
There's a null check two calls in, but I think it'd be good to do it here too. Will fix